### PR TITLE
HDDS-8978. SCM SequenceIdGenerator#reinitialize only needs to acquire the lock once

### PR DESCRIPTION
**Pull request by guohao-rosicky**
## What changes were proposed in this pull request?

SCM SequenceIdGenerator#reinitialize only needs to acquire the lock once

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8978

### CODE REVIEW

**Comment by xichen01** [comment]
Could I ask what are the cases where deadlocks happen, and is this PR https://github.com/apache/ozone/pull/5018 related?

**Comment by Xushaohong** [comment]
The linked Jira is [HDDS-8965], @guohao-rosicky you might need to correct it

**Review by adoroszlai — COMMENTED** [review]
Thanks @guohao-rosicky for working on this.

* Can you please provide more details about the deadlock?  It would be great if you could attach a thread dump to the Jira issue.
* Without knowing the details, I don't think the patch fixes any deadlocks.  `lock` is a `ReentrantLock`.  If it is `lock()`-ed twice, it needs to be `unlock()`-ed twice before it is released.  Thus `lock` is held until return from `reinitialize`, whether it calls `invalidateBatch` or `invalidateBatchInternal`.

**Comment by guohao-rosicky** [comment]
Thanks @adoroszlai  for the comment, as you said, I was wrong in my description, this modification reduces the locks by one since reinitialize only needs to acquire the locks once

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SequenceIdGenerator.java [hunk 1]
```diff
@@ -177,7 +181,7 @@ public void reinitialize(Table<String, Long> sequenceIdTable)
     lock.lock();
     try {
       LOG.info("reinitialize SequenceIdGenerator.");
```
Comment (adoroszlai) [inline hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SequenceIdGenerator.java:183]: I think we should move the log statement out of the lock.

**Review by adoroszlai — COMMENTED** [review]
Thanks @guohao-rosicky for clarifying the intent of this change.

File: hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SequenceIdGenerator.java (hunk 1 above)
Comment (guohao-rosicky) [reply hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/ha/SequenceIdGenerator.java:183]: sure.

**Review by adoroszlai — APPROVED** [review]

**Pull request merged**